### PR TITLE
Fix/clear cluster cache on connection update

### PR DESCRIPTION
**Pull request by Cerchie**
## Summary of Changes

Addresses #142 by clearing cache on connection update and adding test

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

- Tests:
    - [x] Added new
- [x] Have you validated this change locally against a running instance of the Quarkus dev server?
    ```shell
    make quarkus-dev
    ```
- [x] Have you validated this change against a locally running native executable?
    ```shell
    make mvn-package-native && ./target/ide-sidecar-*-runner
    ```

### CODE REVIEW

File: src/test/java/io/confluent/idesidecar/restapi/cache/ClusterCacheTest.java [hunk 1]
```diff
@@ -224,6 +225,16 @@ void shouldFindAlreadyLoadedSchemaRegistryForConnectionAndLazilyLoadKafkaCluster
...
+    @Test
+    void shouldClearCacheOnConnectionUpdateAndReCacheDetails() {
+      // When a connection is updated
+      cache.onConnectionUpdated(CONNECTION_1);
+
+      // assert cache is empty
+      assertEquals(0, cache.clustersByConnectionId.size());
+
+    }
+
```
Comment (rhauch) [inline src/test/java/io/confluent/idesidecar/restapi/cache/ClusterCacheTest.java:237]: I think we essentially want to do something like this:
```suggestion
    @Test
    void shouldRemoveCacheOnConnectionUpdate() {
      // When two connections are created
      cache.onConnectionCreated(CONNECTION_1);
      cache.onConnectionCreated(CONNECTION_2);

      // and a Kafka cluster is loaded in each connection
      cache.onLoadingKafkaCluster(
          new ClusterEvent(CONNECTION_1_ID, ConnectionSpec.ConnectionType.CCLOUD, LKC_1)
      );
      cache.onLoadingKafkaCluster(
          new ClusterEvent(CONNECTION_2_ID, ConnectionSpec.ConnectionType.CCLOUD, LKC_2)
      );

      // Then there will be a connection cache for both connections
      assertConnection(CONNECTION_1);
      assertConnection(CONNECTION_2);

      // and the cache for each connection contains their Kafka cluster
      assertKafkaCluster(LKC_1);
      assertKafkaCluster(LKC_2);

      // When one of the connections is updated
      cache.onConnectionUpdated(CONNECTION_1);

      // then the cache for the connection will have been cleared (no cluster will be found)
      var clusters = assertConnection(CONNECTION_1);
      assertThrows(
          ClusterNotFoundException.class,
          () -> clusters.getKafkaCluster(LKC_1.id(), false)
      );

      // but the cache for the other connection is unaffected
      assertConnection(CONNECTION_2);
      assertKafkaCluster(LKC_2);

      // and the Kafka cluster for connection 1 will be loaded again
      expectFindKafkaCluster(CONNECTION_1_ID, LKC_1.id(), LKC_1);
      // when the Kafka cluster is accessed again
      assertEquals(LKC_1, clusters.getKafkaCluster(LKC_1.id()));
    }
```
This does a couple of things. First, it verifies that the cache for a connection is cleared when it is updated. We're not quite checking that it is cleared, but we're verifying that a cluster that used to be there no longer is in the cache. We could add an `isEmpty()` protected method to the `Clusters` class, and also add an assertion in the test that it is empty.

Second, it verifies that other connections are not affected. The cluster that was loaded into the cache for connection 2 is still there after connection 1 is updated.

Third, it verifies that after a connection is updated and it's cache is cleared (or technically recreated), then the cache for connection 1 will reload the Kafka cluster if it is queried.

BTW, here are the additional constants that are needed for `CONNECTION_2` and `LKC_2`:
```
    static final String CONNECTION_2_ID = "c2";
    static final ConnectionState CONNECTION_2 = ConnectionStates.from(
        new ConnectionSpec(
            CONNECTION_2_ID,
            "Conn2",
            ConnectionSpec.ConnectionType.CCLOUD
        ),
        null
    );
    static final CCloudOrganization ORG_2 = new CCloudOrganization(
        "org2",
        "my-other-org",
        false,
        CONNECTION_2_ID
    );
    static final CCloudEnvironment ENV_2 = new CCloudEnvironment(
        "env-456",
        "my-env-2",
        CCloudGovernancePackage.ESSENTIALS
    ).withOrganization(ORG_2).withConnectionId(CONNECTION_2_ID);
    static final CCloudKafkaCluster LKC_2 = new CCloudKafkaCluster(
        "lkc-2",
        "my-other-kafka-cluster",
        CloudProvider.AWS,
        "us-west-2",
        "pkc-456"
    ).withEnvironment(ENV_2).withOrganization(ORG_2).withConnectionId(CONNECTION_2_ID);
```

File: src/main/java/io/confluent/idesidecar/restapi/cache/ClusterCache.java [hunk 2]
```diff
@@ -191,6 +191,9 @@ void onConnectionCreated(@ObservesAsync @Lifecycle.Created ConnectionState conne
 
   void onConnectionUpdated(@ObservesAsync @Lifecycle.Updated ConnectionState connection) {
     Log.infof("Updated %s", connection.getSpec());
+
+    // clear the cache to force future Kafka/SR REST proxy requests to fetch and re-cache any new cluster details
+    clustersByConnectionId.remove(connection.getSpec().id());
```
Comment (rhauch) [inline src/main/java/io/confluent/idesidecar/restapi/cache/ClusterCache.java:196]: If we just remove it, then when the cache is used to find clusters for a particular connection, that connection will not be found and we'll get an exception.

I think we actually want to _replace_ the connection's cache:
```suggestion

    // Replace the existing cache for this connection with a new one
    clustersByConnectionId.put(
        connection.getId(),
        new Clusters(connection.getId(), connection.getType())
    );
```
BTW, we could also have changed `Clusters` to have a `clear()` method, but that's not as good as _replacing_ the `Clusters` object for the connection. Why? The `Clusters` class has two maps, and it's not possible to atomically clear both. (Another thread might see an inconsistent `Clusters` object between our clearing of the two maps.) However, `clustersByConnectionId` is a concurrent map, and replacing the existing value on that entry requires changing one reference, which is atomic.

**Review by rhauch — COMMENTED** [review]
Thanks, @Cerchie . Based on your questions offline, I've given some suggestions on how to better implement the clearing and test to verify the desired side effects take effect and no undesired side effects take effect. Let me know if this makes sense.

**Review by rhauch — COMMENTED** [review]
Thanks for the changes, @Cerchie. I have a few more suggestions, plus one question for @flippingbits and @rohitsanj below.

File: src/main/java/io/confluent/idesidecar/restapi/cache/ClusterCache.java [hunk 3]
```diff
@@ -2,21 +2,15 @@
...
-import io.confluent.idesidecar.restapi.models.graph.CCloudKafkaCluster;
-import io.confluent.idesidecar.restapi.models.graph.CCloudSchemaRegistry;
-import io.confluent.idesidecar.restapi.models.graph.Cluster;
-import io.confluent.idesidecar.restapi.models.graph.ClusterEvent;
-import io.confluent.idesidecar.restapi.models.graph.KafkaCluster;
-import io.confluent.idesidecar.restapi.models.graph.RealCCloudFetcher;
-import io.confluent.idesidecar.restapi.models.graph.RealDirectFetcher;
-import io.confluent.idesidecar.restapi.models.graph.RealLocalFetcher;
-import io.confluent.idesidecar.restapi.models.graph.SchemaRegistry;
+import io.confluent.idesidecar.restapi.models.graph.*;
```
[the remark this answers is not in the document]
Comment (Cerchie) [reply src/main/java/io/confluent/idesidecar/restapi/cache/ClusterCache.java]: right you had mentioned this in an earlier PR. actually not sure how this one snuck in here, but I'm removing 🧹

File: src/main/java/io/confluent/idesidecar/restapi/messageviewer/MessageViewerClusterInfoProcessor.java [hunk 4]
```diff
@@ -90,6 +90,15 @@ void onConnectionDisconnected(
...
 
+  /**
+   * Respond to the connection being updated by clearing cached information.
+   *
+   * @param connection the connection that was prior to update
+   */
+  void onConnectionUpdated(@ObservesAsync @Lifecycle.Updated ConnectionState connection) {
+    clearCaches();
+  }
+
```
[the remark this answers is not in the document]
Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/messageviewer/MessageViewerClusterInfoProcessor.java:101]: Thanks for catching that. I've created https://github.com/confluentinc/ide-sidecar/issues/153.

[the remark this answers is not in the document]
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/messageviewer/MessageViewerClusterInfoProcessor.java:101]: @Cerchie let's remove the change to this class for now, per the above.

**Review by rhauch — COMMENTED** [review]
Thanks, @Cerchie. Just a few more minor suggestions.

File: src/test/java/io/confluent/idesidecar/restapi/cache/ClusterCacheTest.java [hunk 5]
```diff
@@ -32,6 +14,10 @@
 import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (rhauch) [inline src/test/java/io/confluent/idesidecar/restapi/cache/ClusterCacheTest.java:17]: Please remove the wildcard import.

File: src/test/java/io/confluent/idesidecar/restapi/cache/ClusterCacheTest.java [hunk 6]
```diff
@@ -1,28 +1,11 @@
...
-import io.confluent.idesidecar.restapi.models.graph.CCloudGovernancePackage;
-import io.confluent.idesidecar.restapi.models.graph.CCloudKafkaCluster;
-import io.confluent.idesidecar.restapi.models.graph.CCloudOrganization;
-import io.confluent.idesidecar.restapi.models.graph.CCloudSchemaRegistry;
-import io.confluent.idesidecar.restapi.models.graph.CloudProvider;
-import io.confluent.idesidecar.restapi.models.graph.ClusterEvent;
-import io.confluent.idesidecar.restapi.models.graph.KafkaCluster;
-import io.confluent.idesidecar.restapi.models.graph.RealCCloudFetcher;
-import io.confluent.idesidecar.restapi.models.graph.RealLocalFetcher;
+import io.confluent.idesidecar.restapi.models.graph.*;
```
Comment (rhauch) [inline src/test/java/io/confluent/idesidecar/restapi/cache/ClusterCacheTest.java:8]: Please remove the wildcard import.

**Review by rhauch — COMMENTED** [review]
Thanks, @Cerchie. Two more wildcard imports need to be removed, but otherwise looks great.

File: src/test/java/io/confluent/idesidecar/restapi/cache/ClusterCacheTest.java (hunk 6 above)
Comment (Cerchie) [reply src/test/java/io/confluent/idesidecar/restapi/cache/ClusterCacheTest.java:8]: removing! also did [this ](https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#disable-wildcard-imports)so these shouldn't appear anymore

**Review by rhauch — APPROVED** [review]
Very nice work, @Cerchie!
